classes/local.php

```diff
@@ -266,31 +266,33 @@ public static function file_pluginfile($relativepath) {
         file_pluginfile($relativepath, $forcedownload, $preview, $offline, $embed);
     }
 
-    public static function file_is_public(stored_file $file): bool {
+    /**
+     * Returns true if a file is considered public.
+     *
+     * @param string $contenthash
+     * @return bool
+     */
+    public static function file_is_public(string $contenthash): bool {
         global $DB;
 
         $minduplicates = get_config('filter_imageopt', 'minduplicates');
         $publicfilescache = \cache::make('filter_imageopt', 'public_files');
-        $key = 'public_files_' . $minduplicates;
+        $key = 'publicfiles' . '_' . $contenthash . '_' . $minduplicates;
```

Comment:
Hey Thomas, no worries. You're right about there being potentially many DB hits before the cache is all built.

[Other parts of this PR](https://github.com/gthomas2/moodle-filter_imageopt/pull/41/commits/0743302f8b78b317161bae95ccfd13889f45103a) reduce the number of DB hits per image. I suspect that due caching the processed tags, the net number of DB queries is now less, even with the one query added per image for duplicate detection. In the numbers above, with duplicate detection disabled, the number of DB queries went from 46->121 when the vanilla plugin was enabled. With the processed tags cache (and duplicate detection still off) that goes back down to 46. But I'd need to run tests again to be sure the net number of queries is still the same or less overall.

I considered the scheduled task idea, but the files table on the site where we're testing this has over one hundred million rows. The original SQL query takes about 6 minutes to finish (even when minduplicates is set to something like 500) and the resulting array can be enormous if minduplicates is set to something low.

We could possibly do something smart in the task like cache an array of hashes per context id, and process more recently accessed courses first, but the current patch sort of does that already since it adds items to the cache based on what is actually being accessed (i.e., a course that is accessed frequently will have its images processed and put in the cache, whereas a course that is never accessed will not).

I guess there's pros and cons to both approaches. I'm not sure which is the best. Happy to be steered.

Reply:
I need to put some in depth thought into this and real world shaped load testing. But my knee jerk reaction is that the granularity of the cache keys should be roughly one per page view. So no idea if this is viable but could we rejig things that that we build up a structure that contains a map of multiple old urls to new urls and it is stored at a fairly chunky level. I'm thinking if you have a course page with 10 mod_label's on it then they all end up as one key. We'd need to do some level of url processing so that it included key parameters in the key, but excluded ones which were not important and that logic might be funky. Also open to other suggestions like the page's context id could be the key.

Perhaps an more crazy idea: We can also set limits to how much slower the page is allowed to be while generating this stuff, ie if we have a huge page with 1000 images then the filter could be smart enough to stop processing after 50 lookups and saving its results in the cache. And then over 20 page views the page would be fully warmed up.

Reply:
The way this branch currently works is kind of already storing a mapping between old URLs and new URLs. Once a tag has been processed, the [result of that is cached](https://github.com/gthomas2/moodle-filter_imageopt/pull/41/commits/0743302f8b78b317161bae95ccfd13889f45103a#diff-7e65ffe11d9b1dec1ce0329e2917c0782e7ac4a2b3c1dc6dfe995088fe66f763R234) and then used when we encounter that URL again. The public files cache won't be used again for that URL after that point. Currently the key is [generated from the input img tag](https://github.com/gthomas2/moodle-filter_imageopt/pull/41/commits/0743302f8b78b317161bae95ccfd13889f45103a#diff-7e65ffe11d9b1dec1ce0329e2917c0782e7ac4a2b3c1dc6dfe995088fe66f763R179) and the value is the URL of the processed image, but I'm pretty sure that can be changed so that the key is a contextid (or something similar) and the value is a mapping between raw URLs and processed URLs.

Before processing an img tag, we need to know if the source image is public or private. I think we could remove the public files cache completely and the end result would be exactly the same. But having the public files cache means we use less DB queries as we're filling up the processed tags cache. If it doesn't seem worth it, then we could remove it.

What you mention in the second paragraph is kind of already happening (that may only be because of the loadonvisible setting though). In my testing it took something like 4 or 5 refreshes for the cache to be fully populated.